include/rcutils/filesystem.h

```diff
@@ -185,19 +192,49 @@ bool
 rcutils_mkdir(const char * abs_path);
 
 /// Calculate the size of the specified directory.
-/*
+/**
  * Calculates the size of a directory by summarizing the file size of all files.
  * \note This operation is not recursive.
  * \param[in] directory_path The directory path to calculate the size of.
  * \param[in] allocator Allocator being used for internal file path composition.
- * \return The size of the directory in bytes.
+ * \return The size of the directory in bytes on success.
  */
 RCUTILS_PUBLIC
 size_t
 rcutils_calculate_directory_size(const char * directory_path, rcutils_allocator_t allocator);
 
+/// Calculate the size of the specified directory with recursion.
+/**
+ * Calculates the size of a directory and subdirectory by summarizing the file size of all files.
+ * If necessary, you can specify the maximum directory depth to recurse into.
+ * Depth definition as below.
+ * \code
+ * directory_path  <= depth 1
+ *    |- subdirectory <= depth 2
+ *            |- subdirectory <= depth 3
+ *                    ...
+ * \endcode
+ *
+ * \note Not calculate the size of the symlink which points to file or directory.
+ * \param[in] directory_path The directory path to calculate the size of.
+ * \param[in] max_depth The maximum depth of subdirectory. 0 means no limitation.
+ * \param[out] size The size of the directory in bytes on success
+ * \param[in] allocator Allocator being used for internal file path composition.
+ * \return `RCUTILS_RET_OK` if successful, or
+ * \return `RCUTILS_RET_INVALID_ARGUMENT` for invalid arguments, or
+ * \return `RCUTILS_RET_BAD_ALLOC` if memory allocation fails
+ * \return `RCUTILS_RET_ERROR` if other error occurs
+ */
+RCUTILS_PUBLIC
+rcutils_ret_t
+rcutils_calculate_directory_size_with_recursion(
+  const char * directory_path,
+  const size_t max_depth,
+  size_t * size,
```

Comment:
I don't think this works on 32-bit; as far as I know, this is defined as `unsigned int`, and hence will still be 32-bit there.  I think we should explicitly use `uint64_t` here.

Reply:
> I don't think this works on 32-bit; as far as I know, this is defined as unsigned int, and hence will still be 32-bit there. I think we should explicitly use uint64_t here.

Yeah.  It's my fault.  
I understand why your suggestion use uint64_t instead of size_t.

Reply:
The declaration here still needs to be changes to `uint64_t`.